### PR TITLE
Consider arch when looking for compatible dist.

### PR DESCRIPTION
**Pull request by Aralox**
Cache self.dist so it isn't calculated multiple times (e.g. wrapper_func in toolchain).

Current problem scenario:
- I build for arm
- I build the same project for x86. p4a uses existing distribution and packages arm libraries into the apk instead of recompiling project for x86

I also modified _dist() as it was getting called multiple times, causing the dist search (and log output) twice.

### CODE REVIEW

File: pythonforandroid/toolchain.py [hunk 1]
```diff
@@ -792,9 +793,9 @@ def export_dist(self, args):
 
     @property
     def _dist(self):
-        ctx = self.ctx
-        dist = dist_from_args(ctx, self.args)
-        return dist
+        if (not hasattr(self, 'dist') or self.dist is None):
+            self.dist = dist_from_args(self.ctx, self.args)
+        return self.dist
```
Comment (AndreMiras) [inline pythonforandroid/toolchain.py]: I have to say this feels unnatural somehow:
1) I would simplify the initialisation of the attribute by assigning it to `None` within the constructor
2) It feels weird to have the property being "self._protected" while the member "self.public". Maybe we want to call the member  `__dist` at least? I don't know

File: pythonforandroid/distribution.py [hunk 2]
```diff
@@ -92,15 +95,13 @@ def get_distribution(cls, ctx, name=None, recipes=[],
...
-            ) or dist.ndk_api is None:
+            if ((ndk_api is not None and dist.ndk_api != ndk_api)
+                or dist.ndk_api is None or dist.archs is None):
                 continue
-            for recipe in recipes:
-                if recipe not in dist.recipes:
-                    break
-            else:
+            if (set(recipes).issubset(set(dist.recipes))
+                and set(archs).issubset(set(dist.archs))):
```
Comment (AndreMiras) [inline pythonforandroid/distribution.py]: Linter was complaining about these lines:
```
pythonforandroid/distribution.py:99:5: E129 visually indented line with same indent as next logical line
pythonforandroid/distribution.py:102:17: E129 visually indented line with same indent as next logical line
```
Could you fix so Travis finish the build in order to check eventual regressions?

**Comment by AndreMiras** [comment]
Thank you for your contribution :beer: 
I didn't get time to really wrap my head around it and play with it yet. Just left a couple of comments

**Comment by Aralox** [comment]
This pr isn't urgent. Workaround is to run `p4a clean_dists` between compiling for different architectures. I think the other pr currently in progress for upgrading to ndk 19 might already take this into account.

**Comment by AndreMiras** [comment]
Hi @Aralox is it still relevant after changes from https://github.com/kivy/python-for-android/pull/1722 ?
If I understood correctly, we can verify this by compiling for two different archs without running a clean between and checking if the APKs run OK on both. If the issue is still here, the second arch we build for would have compiled dists from the previous one hence should crash, correct?

**Comment by Aralox** [comment]
Hi Andre yes I think that's exactly it. I haven't really kept up with the progress of this project so I'm not sure if the other PR addresses this (it probably does), but yeah that would be the verification. Feel free to close this PR if unnecessary. Cheers :)

**Comment by inclement** [comment]
@AndreMiras As far as I can see, #1722 doesn't affect this and it's still an issue, am I missing something?

**Comment by opacam** [comment]
Yeah, agree with @inclement, I don't see any conflict with the NDK-r19 migration thing...but I just now realize that this will probably conflict with #1926 (as the author of this last one, no problem at all, I will fix conflicts if convenient :wink:).

**Comment by inclement** [comment]
I'm dealing with this in a new PR.

**Comment by opacam** [comment]
@inclement, good to know :+1:, feel free to use the commits I made in  at #1926, related with this arch problem, I see two:
  - [Check android_api and archs when reusing dist](https://github.com/kivy/python-for-android/pull/1926/commits/1e7d0d4112606b30bae1674f3a8352837b558358)
  - [Add an arch suffix to our dists](https://github.com/kivy/python-for-android/pull/1926/commits/096e3e5874ba5ff908b3c705270eb26112e6b97c)

Of course, feel free to implement it whatever way you want, you may have a better idea :smile:

**Comment by inclement** [comment]
@opacam I'm feeling inclined to make a simpler PR that addresses just this for now, in a slightly different way, with #1926 left to merge for the next release. I don't want to merge #1926 quite yet because I'm concerned it may be a little overcomplicated.

**Comment by opacam** [comment]
@inclement, no problem at all, I agree that #1926 complicates things... if we can slim it in some way better for all of us :wink:

so go for this new PR :+1:!

**Comment by inclement** [comment]
Closing as I think this was resolved.

**Pull request closed without merging**